specs/lightclient/validator.md

```diff
@@ -0,0 +1,383 @@
+
+Selection proofs are provided in `ContributionAndProof` to prove to the gossip channel that the validator has been selected as an aggregator.
+
+`ContributionAndProof` messages are signed by the aggregator and broadcast inside of `SignedContributionAndProof` objects to prevent a class of DoS attacks and message forgeries.
+
+First, `contribution_and_proof = get_contribution_and_proof(state, validator_index, contribution, privkey)` is constructed.
+
+```python
+def get_contribution_and_proof(state: BeaconState,
+                               aggregator_index: ValidatorIndex,
+                               contribution: SyncCommitteeContribution,
+                               privkey: int) -> ContributionAndProof:
+    selection_proof = get_sync_committee_slot_signature(
+      state, 
+      contribution.slot,
+      privkey
+    )
+    return ContributionAndProof(
+        aggregator_index=aggregator_index,
+        contribution=contribution,
+        selection_proof=selection_proof,
+    )
+```
+
+Then `signed_contribution_and_proof = SignedContributionAndProof(message=contribution_and_proof, signature=signature)` is constructed and broadast. Where `signature` is obtained from:
+
+```python
+def get_contribution_and_proof_signature(
+  state: BeaconState,
+  contribution_and_proof: ContributionAndProof,
+  privkey: int) -> BLSSignature:
+    contribution = contribution_and_proof.contribution
+    domain = get_domain(state, DOMAIN_CONTRIBUTION_AND_PROOF, compute_epoch_at_slot(contribution.slot))
+    signing_root = compute_signing_root(contribution_and_proof, domain)
+    return bls.Sign(privkey, signing_root)
+```
+
+## Sync committee subnet stability
+
+The sync committee subnets need special care to ensure stability given the relatively low number of validators involved in the sync committee at any particular time. To provide this stability, a validator must do the following:
+
+* Maintain advertisement of the subnet the validator in the sync committee is assigned to in their node's ENR as soon as they have joined the subnet. Subnet assignments are known `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` epochs in advance and can be computed with `compute_subnet_for_sync_committee` defined above. ENR advertisement is indicated by setting the `syncnets` entry in the ENR to a `uint8` encoding of the `subnet_id`, e.g. `ENR["syncnets"] = b'\x07'`. This entry is removed from the ENR after the validator has left the sync committee.
+
+- TODO: what happens for the first sync committee?
+
+* To join a sync committee subnet, select a random number of epochs before the end of the current sync committee period between 1 and `SYNC_COMMITTEE_SUBNET_COUNT`, inclusive. Validators should join their member subnet at the beginning of the epoch they have randomly selected. For example, if the next sync committee period starts at epoch `853,248` and the validator randomly selects an offset of `3`, they should join the subnet at the beginning of epoch `853,245`.
```

Comment:
the thinking here is to smooth out the "sharpness" of the transition of having one chunk of the sync committee suddenly join a subnet. an alternative that may actually be fine (and is simpler to implement) is to just have the entire next sync committee join their respective subnet at a fixed epoch ahead of time, e.g. `boundary_epoch` - 8 epochs